plugins/BEdita/Core/src/Model/Action/SignupUserAction.php

```diff
@@ -174,17 +192,89 @@ protected function createUser(array $data)
+
+        $user = $this->createUserEntity($data, 'on', 'signupExternal');
+
+        // create `ExternalAuth` entry
+        $this->Users->dispatchEvent('Auth.externalAuth', [
+            'authProvider' => $authProvider,
+            'providerUsername' => $data['provider_username'],
+            'userId' => $user->get('id'),
+            'params' => empty($data['provider_userdata']) ? null : $data['provider_userdata'],
+        ]);
+
+        return $user;
+    }
+
+    /**
+     * Create User model entity.
+     *
+     * @param array $data The signup data
+     * @param string $status User `status`, `on` or `draft`
+     * @param string $validate Validation options to use
+     * @return @return \BEdita\Core\Model\Entity\User The User entity created
+     */
+    protected function createUserEntity(array $data, $status, $validate)
+    {
         $action = new SaveEntityAction(['table' => $this->Users]);
 
         return $action([
             'entity' => $this->Users->newEntity()->set('status', $status),
             'data' => $data,
             'entityOptions' => [
-                'validate' => 'signup',
+                'validate' => $validate,
             ],
         ]);
     }
 
+    /**
+     * Check external auth data validity
+     *
+     * To perform external auth check these fields are mandatory:
+     *  - "auth_provider": provider name like "google", "facebook"... must be in `auth_providers`
+     *  - "provider_username": id or username of the provider
+     *  - "access_token": token returned by provider to use in check
+     *
+     * @param array $data The signup data
+     * @return \BEdita\Core\Model\Entity\AuthProvider AuthProvider entity
+     * @throws \Cake\Network\Exception\UnauthorizedException Upon external authorization check failure.
+     */
+    protected function checkExternalAuth(array $data)
+    {
+        /** @var \BEdita\Core\Model\Entity\AuthProvider $authProvider */
+        $authProvider = TableRegistry::get('AuthProviders')->find('enabled', ['name' => $data['auth_provider']])->first();
+        if (empty($authProvider)) {
+            throw new UnauthorizedException(__d('bedita', 'External auth provider not found'));
+        }
+        $providerResponse = $this->getOAuth2Response($authProvider->get('url'), $data['access_token']);
+        if (!$authProvider->checkAuthorization($providerResponse, $data['provider_username'])) {
+            throw new UnauthorizedException(__d('bedita', 'External auth failed'));
+        }
+
+        return $authProvider;
+    }
+
+    /**
+     * Get response from an OAuth2 provider
+     *
+     * @param string $url OAuth2 provider URL
+     * @param string $accessToken Access token to use in request
+     * @return array Response from an OAuth2 provider
+     * @codeCoverageIgnore
+     */
+    protected function getOAuth2Response($url, $accessToken)
+    {
+        /** @var \Cake\Http\Client\Response $response */
```

Comment:
[minor] Is this type annotation required? 🤔

Reply:
not here indeed...